### DISCUSSION TITLE
Location of wkhtmltopdf

### DISCUSSION THREAD

**Discussion by mrajalko** [serenity-is/Serenity#5654] [Help] [answered]
Please help me in placement of wkhtmltopdf.exe in the project so that it is copied during the publish also.

**Comment by minhhungit** [comment]
Copy the file to root of your project, right click on it, set copy if newer 

After building,the file will be copied to bin folder

**Reply by minhhungit** [reply]
![image](https://user-images.githubusercontent.com/2279508/115863422-9c88ee00-a45f-11eb-92d6-1af37bb75ce1.png)

**Reply by mrajalko** [reply]
It Worked.

but I made a few changes as under

in place of line
var wkhtmlPath = System.IO.Path.Combine(typeof(ReportController).Assembly.Location, "wkhtmltopdf.exe");

I replaced with this line.

var wkhtmlPath = System.IO.Path.Combine(System.IO.Path.GetDirectoryName(typeof(ReportController).Assembly.Location), "wkhtmltopdf.exe");
![Serenity Path Issue](https://user-images.githubusercontent.com/18166582/115959557-66fd0700-a52a-11eb-98e7-a35a263d3534.png)
![Serenity Path Issue -1](https://user-images.githubusercontent.com/18166582/115959561-682e3400-a52a-11eb-81d7-26b82406f0aa.png)
![Serenity Path Issue -2](https://user-images.githubusercontent.com/18166582/115959562-68c6ca80-a52a-11eb-9e16-cf4cb9dcca03.png)

**Comment by mrajalko** [comment]
Surprisingly it works in development environment but doesn't work when published.

**Comment by obaidalrushdi** [comment]
Hi 
copy the file to wwwroot and use this 
 var wkhtmlPath= System.IO.Path.Combine(HostEnvironment.WebRootPath, "wkhtmltopdf.exe");

**Reply by mrajalko** [reply]
Thanks for your suggestion, but the effect remains same, this solution also only woks in development environment and throws error after publish in production environment. may be I am missing something very minor or making a silly mistake somewhere but I am unable to trace where.
![Report Render](https://user-images.githubusercontent.com/18166582/116847821-f7150d80-ac08-11eb-8a80-bab658eb5298.png)

**Reply by mrajalko** [reply]
Request someone to please suggest how to trace values of variables in report controller in browser console.

**Comment by eissa1005** [comment]
@mrajalko 
Try this Code
 string PDFPath = string.Format(@"{0}\{1}", Environment.CurrentDirectory, "wkhtmltopdf.exe");
            var converter = new HtmlToPdfConverter();
            var wkhtmlPath = Path.Combine(typeof(ReportController).Assembly.Location, PDFPath);
            if (!System.IO.File.Exists(wkhtmlPath)) // linux?
                wkhtmlPath = Path.Combine(typeof(ReportController).Assembly.Location, "wkhtmltopdf");

            if (System.IO.File.Exists(wkhtmlPath))
                converter.UtilityExePath = wkhtmlPath;

**Reply by mrajalko** [reply]
Not working.
![image](https://user-images.githubusercontent.com/18166582/121640619-c5b31b80-caab-11eb-8399-d06b0e047209.png)

and

![image](https://user-images.githubusercontent.com/18166582/121640779-f98e4100-caab-11eb-9ca6-c941435ff2cf.png)

**Comment by VictorTomaili** [comment, marked as answer]
Just add an appsettings configuration for a path of wkhtmltopdf and read it from ReportController. @mrajalko

**Reply by mrajalko** [reply]
thanks @VictorTomaili  I would try that.

**Reply by mrajalko** [reply]
implemented the code change as per your suggestion, but the situation remains same it works in development environment but fails in production.

can you please suggest how I display the value of some variables in console.log from ReportController so I may find the cause for such behavior.

Incidentally the machine is same for both development and production environment.

thanks @VictorTomaili for your support.

**Reply by VictorTomaili** [reply]
With an `appsettings` path like `c:\yourapp\wkhtmltopdf.exe` should work. There is no reason for `FileNotFoundException` etc. 

You can access exceptions under [Administration/ExceptionLog](https://serenity.is/demo/Administration/ExceptionLog/) menu. Firstly you need to check this exception page to determine the actual exception you faced. 

It's seems something different exception occurred than path exception. In production, your executable runs well, but your report's output is `null`, and it can't be downloaded because of `null` `fileContents`. https://github.com/serenity-is/Serenity/discussions/5654#discussioncomment-686630

You can add logs by injecting `IExceptionLogger` from the constructor. Log your file path with it. I ensure you will find the exact path of your `wkhtmltopdf` executable when you logged.

**Reply by mrajalko** [reply]
Thanks @VictorTomaili and I agree that the error might be because report output is null.

but the question now is with same data and same line of code in development environment is sending the html to wkhtmltppdf tool but not during the production.

I would use IExceptionLogger as suggested by you and revert with my findings on the same.

**Reply by VictorTomaili** [reply]
Maybe data differences? Do you use the same database in production and development?

**Reply by mrajalko** [reply]
Yes data is same
same connection string is used.

**Reply by VictorTomaili** [reply]
Something is different in release mode from debug mode. It's not about wkhtmltopdf path or serenity related. Something about inside your report controller. Maybe the release path contains a different version of wkhtmltopdf executable. Or some DLLs are not changed when you published because they are already used from another process etc. Or your IIS has configured for the old application path which gives this error and you miss this configuration. Or you have some different environment configuration for the application. Or you have some condition like `#IF Debug`. I can't do a blind guess. Too many possibilities exist. Try to add step-by-step logs and find the exact point of interest.

**Reply by mrajalko** [reply]
Yes I agree @VictorTomaili 

I would use IExcptionLogger and fid out what is the exact issue and report in this forum.

thanks for your help

**Reply by mrajalko** [reply]
hi @VictorTomaili please share a code snippet on how to use IExceptionLogger inside report controller to view values of variables in console.log in production environment.

**Reply by VictorTomaili** [reply]
Retrieve it from dependency injection. Just like IUserAccessor and IPermissionService etc.

**Reply by mrajalko** [reply]
I have started with defining the logger as 
![Definition](https://user-images.githubusercontent.com/18166582/122195431-faf8a880-ceb3-11eb-9027-4d69cb8f8410.png)

and then I have initialized the value as
![Intialization](https://user-images.githubusercontent.com/18166582/122195546-182d7700-ceb4-11eb-9677-48a9f54308b5.png)


and then in the renderasPDF make the function in try catch to catch the exception as

![RenderAsPdf-1](https://user-images.githubusercontent.com/18166582/122195779-490dac00-ceb4-11eb-9cdc-32e20537744d.png)


and in catch as
![RenderAsPdf-2](https://user-images.githubusercontent.com/18166582/122195857-575bc800-ceb4-11eb-9e72-8fd6d4cee912.png)

**Reply by mrajalko** [reply]
have I understood your suggestion correctly.

**Reply by VictorTomaili** [reply]
If you see logs on the "Exception Logs" page, yeah.

**Reply by mrajalko** [reply]
But it is only available for StartSharp Users ??

is there any way by which I can see values of variables on report controller in console.log

could the cause of this behavior some browser setting.

**Reply by VictorTomaili** [reply]
Oh, I see. You are right It's a pro feature on startsharp. You can implement your logger as you wish with this interface. Or you can implement any other logger like serilog etc.

**Reply by mrajalko** [reply]
Finally found the issue. it was related to host name configuration on my publishing server.

Thanks @VictorTomaili for your help.

**Comment by imhemayatsangin** [comment]
Hello dear brother, please help me out im stuck in this since a month i tried all possible ways and i really don't know where im doing mistakes. I donwloaded the wkhtmltopdf.exe and installed it normally on my machine, then i coppied the wkhtmltopdf.exe file which was located in the C:\Program files and copied to the root folder of the project and still it's giving me the error, even i downloaded the wktmltopdf.exe file and without installation i copied to the root directory but none of the way it's working so please guide me what should i do? i'm using asp.net core version of the Serene. If you want me i can send you my project file so that anybody can help me.

**Reply by mrajalko** [reply]
Are you facing issue on development machine or on production.

please share your project on google drive and share on my email mrajalko@gmail.com 
Let me see if I can help.

**Reply by imhemayatsangin** [reply]
Thanks Brother for the prompt response! Surprisingly it worked, But still if i get any issues then i will ping you via email brother. Thank you so much.
